### DISCUSSION TITLE
Using Optiboot with Power Line Communication

### DISCUSSION THREAD

**Discussion by farfarfromhome** [Optiboot/optiboot#324] [General]
I am using an attiny1614 as the controller in a Swimming Pool Light.  The only connection to the light is 12VAC and I am planning to use PLC (Power Line Communication) to communicate to the chip.  The interface to the PLC is half duplex, 1200 baud serial, so it looks like a slow speed RS485 interface.  The time to send my typical program to the attiny1614 would be less that 1 minute.  Hence using Optiboot is much more attractive than lowering the swimming pool water, unscrewing the module, re-program, re-fit, and re-fill the swimming pool!

I read that RS485 is already supported via enable/disable transmitter so is Optiboot communication inherently half duplex?  Can I just connect to a serial port at 1200 baud?

I am also planning to use an ESP32 controller to communicate to several of these lights, so is it possible to set a communication relay thru the ESP32, ideally wifi to serial port?  Comments on feasibility, difficulty level, and hints of how to do this would be greatly appreciated.

**Comment by WestfW** [comment]
> is Optiboot communication inherently half duplex?

I believe it is.  the STK500v1 protocol is very (entirely?) command/response based - the PC sends a command, which the bootloader parses and sends a response.  This should work fine on a half-duplex link, if the link is "turned around" at the appropriate places.

It looks to me like the current rs485 support does NOT do a very good job of handling this turn-around.  Instead of switching on a per-command basis, it switches on a per-character basis, which means it's pretty dependent on the external circuitry being able to do the switch pretty much instantly.

**Comment by DRSDavidSoft** [comment]
@farfarfromhome I'm really invested in what you used in the end to achieve your goal, similarly, I'm designing a board that coincidentally is also used as a swimming pool light. I'm lucky enough that I can run any cables I need, including ones for data & communication, so no need for a PLC in my case, but I'm interested to see what you were able to achieve at the end in your case!
